aws/s2n-tls
# fix: close all /dev/urandom open fds

### Resolved issues:

Partially solve https://github.com/aws/s2n-tls/issues/4005

### Description of changes: 

* Directly call `s2n_rand_cleanup` to close `/dev/urandom` in `s2n_drbg_test.c` and `s2n_override_openssl_random_test.c`.
    * `s2n_drbg_test.c` and `s2n_override_openssl_random_test.c` calls `s2n_rand_set_callbacks` function which resets the default cleanup callback functions. 
    * The `rand_cleanup_callback` parameter is supposed to close `/dev/urandom`, but `s2n_cleanup` no longer calls that function, since that parameter is overwritten to some other clean up functions, like `nist_fake_entropy_init_cleanup`.
    * Hence, I call `s2n_rand_cleanup` to run those fake clean up call back functions, and reset those call back functions back to default.
    * In that way, `END_TEST` will cleanup the default callbacks set by `BEGIN_TEST`, so that the default callbacks won't leak fds.
    * The solution is to call `s2n_rand_cleanup` to cleanup the default random callbacks. Then we call `s2n_rand_init` to reinitiate the randomness engine again. 
* Some other options I tried for the previous problem
    * Implement another [possible fix](https://github.com/aws/s2n-tls/pull/4835#discussion_r1795664188). Look for that comment for reasons why it wouldn't work.
    * Call `s2n_rand_set_callbacks` before `BEGIN_TEST`.
        * `s2n_drbg_test.c` wants to pass two different generators to `s2n_rand_set_callbacks`, so fix a generator before `BEGIN_TEST` will change the test itself.
* Close `/dev/urandom` in `s2n_fork_generation_number_test.c`.
    * Call `s2n_rand_clean_up()` directly for the reason of [this comment](https://github.com/aws/s2n-tls/pull/4835#discussion_r1802028152).   
* Close `dev_urandom` in `s2n_random_test.c` for both test 0 and 1.
    * For test 1, we need to close the original fd for `dev_urandom` before we make that fd invalid.

### Call-outs:

* See [PR#4833](https://github.com/aws/s2n-tls/pull/4833) and [PR#4834](https://github.com/aws/s2n-tls/pull/4834) for fixes of other open fds.
* I didn't add a test in this PR. The test to detect all opened fds will be in a separate PR.

### Testing:

* Test locally. 
    * Add `--track-fds=yes` to CTest memcheck and direct all Valgrind output to `MemoryTester.log` files.
    * Search in VS Code for Open File Descriptors of `/dev/urandom`.
        * Previously opened `/dev/urandom` in tests that are fixed.
    * This is similar to the check in `./codebuild/bin/test_exec_leak.sh`.
  https://github.com/aws/s2n-tls/blob/6bb195c23fd1ecf623dece3b43644e7c9a9e0e20/codebuild/bin/test_exec_leak.sh#L82-L97
* Open fds test will be implemented once [PR#4851](https://github.com/aws/s2n-tls/pull/4851) is merged.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
